References
----------

.. [Bass13] J. I. F. Bass, A. Diallo, J. Nelson, J. M. Soto, C. L. Myers, and A. J. Walhout,
   "Using networks to measure similarity between genes: association index selection",
   In `Nature methods <https://www.nature.com/articles/nmeth.2728>`_, 2013.

.. [Belkina19] A. C. Belkina, C. O. Ciccolella, R. Anno, R. Halpert, J. Spidlen, and J. E. Snyder-Cappione,
   "Automated optimized parameters for T-distributed stochastic neighbor embedding improve visualization and analysis of large datasets",
   In `Nature Communications <https://www.nature.com/articles/s41467-019-13055-y>`__, 2019.

.. [Blondel08] V. D. Blondel, J.-L. Guillaume, R. Lambiotte, and E. Lefebvre,
   "Fast unfolding of communities in large networks",
   In `Journal of Statistical Mechanics: Theory and Experiment <https://iopscience.iop.org/article/10.1088/1742-5468/2008/10/P10008/meta>`_, 2008.

.. [Büttner18] M. Büttner, et al.,
   "A test metric for assessing single-cell RNA-seq batch correction",
   In `Nature Methods <https://www.nature.com/articles/s41592-018-0254-1>`__, 2018.

.. [Gale95] W. A. Gale, and G. Sampson.
   "Good‐turing frequency estimation without tears.",
   In `Journal of quantitative linguistics <https://www.grsampson.net/AGTf.pdf>`__, 1995.

.. [Gao21] Ch. Gao, et al.,
   "Iterative single-cell multi-omic integration using online learning",
   In `Nature Biotechnology <https://www.nature.com/articles/s41587-021-00867-x>`__, 2021.

.. [Hie19] B. Hie, B. Bryson, and B. Berger,
   "Efficient integration of heterogeneous single-cell transcriptomes using Scanorama",
   In `Nature Biotechnology <https://www.nature.com/articles/s41587-019-0113-3>`__, 2019.

.. [Jacomy14] M. Jacomy, T. Venturini, S. Heymann, and M. Bastian,
   "ForceAtlas2, a continuous graph layout algorithm for handy network visualization designed for the Gephi software",
   In `PLoS ONE <https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0098679>`__, 2014.

.. [Kobak19] D. Kobak, and P. Berens,
   "Kobak, D., & Berens, P. (2019). The art of using t-SNE for single-cell transcriptomics",
   In `Nature Communications <https://www.nature.com/articles/s41467-019-13056-x>`__, 2019.

.. [Korsunsky19] I. Korsunsky, et al.,
   "Fast, sensitive and accurate integration of single-cell data with Harmony",
   In `Nature Methods <https://www.nature.com/articles/s41592-019-0619-0>`__, 2019.

.. [Li20] B. Li, J. Gould, Y. Yang, et al.,
   "Cumulus provides cloud-based data analysis for large-scale single-cell and single-nucleus RNA-seq",
   In `Nature Methods <https://www.nature.com/articles/s41592-020-0905-x>`__, 2020.

.. [Li20-1] B. Li,
   "A streamlined method for signature score calculation",
   In `Pegasus repository <https://github.com/lilab-bcb/pegasus/raw/master/signature_score.pdf>`__, 2020.

.. [Li20-2] B. Li,
   "Doublet detection in Pegasus",
   In `Pegasus repository <https://github.com/lilab-bcb/pegasus/raw/master/doublet_detection.pdf>`__, 2020.

.. [Linderman19] G.C. Linderman, et al.,
   "Fast interpolation-based t-SNE for improved visualization of single-cell RNA-seq data",
   In `Nature Methods <https://www.nature.com/articles/s41592-018-0308-4>`__, 2019.

.. [Jerby-Arnon18] L. Jerby-Arnon, et al.,
   "A cancer cell program promotes T cell exclusion and resistance to checkpoint blockade",
   In `Cell <https://www.sciencedirect.com/science/article/pii/S0092867418311784>`__, 2018.

.. [Malkov16] Yu. A. Malkov, and D.A. Yashunin,
   "Efficient and robust approximate nearest neighbor search using Hierarchical Navigable Small World graphs",
   In `arXiv <https://arxiv.org/abs/1603.09320>`__, 2016.

.. [McInnes18] L. McInnes, J. Healy, and J. Melville,
   "UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction",
   Preprint at `arXiv <https://arxiv.org/abs/1802.03426>`__, 2018.

.. [Suo18] S. Suo, Q. Zhu, A. Saadatpour, L. Fei, G. Guo, and G. Yuan,
   "Revealing the critical regulators of cell identity in the mouse cell atlas",
   In `Cell reports <https://www.sciencedirect.com/science/article/pii/S2211124718316346>`_, 2018.

.. [Traag19] V. A. Traag, L. Waltman, and N. J. van Eck,
   "From Louvain to Leiden: guaranteeing well-connected communities",
   In `Scientific Reports <https://www.nature.com/articles/s41598-019-41695-z>`__, 2019.

.. [Welch19] J. D. Welch, et al.,
   "Single-Cell Multi-omic Integration Compares and Contrasts Features of Brain Cell Identity"
   In `Cell <https://www.sciencedirect.com/science/article/pii/S0092867419305045>`__, 2019.

.. [Wolock18] S.L. Wolock, R. Lopez, and A.M. Klein,
   "Scrublet: computational identification of cell doublets in single-cell transcriptomic Data",
   In `Cell Systems <https://www.sciencedirect.com/science/article/pii/S2405471218304745>`__, 2018.

.. [Yang16] Z. Yang, and G. Michailidis,
   "A non-negative matrix factorization method for detecting modules in heterogeneous omics multi-modal data",
   In `Bioinformatics <https://academic.oup.com/bioinformatics/article/32/1/1/1743821?login=true>`__, 2016.
